GH-118036: Fix a bug with CALL_STAT_INC

gvanrossum opened the pull request:
We were under-counting calls in `_PyEvalFramePushAndInit` because the `CALL_STAT_INC` macro was redefined to a no-op for the Tier 2 interpreter. The fix is a little convoluted (I had wanted to move the code around, but that would require moving something else around, and in the end I figured it was easier to tweak the macros @markshannon might disagree though?). This ought to result in ~37% more "Frames pushed" reported under "Call stats". The new count is the correct one (I presume).

@mdboom can you review? This is [one commit](https://github.com/python/cpython/pull/117908/commits/0c195cb8a6af4cdfe382013602eb5d256ddcb301) from my experiment about removing Tier 2 entirely (gh-117908).

To see the effect, look at [these pystat diffs](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240415-3.13.0a6%2B-2055e5f/bm-20240415-azure-x86_64-gvanrossum-disable_tier2-3.13.0a6%2B-2055e5f-pystats-vs-base.md#call-stats).

<!-- gh-issue-number: gh-118036 -->
* Issue: gh-118036
<!-- /gh-issue-number -->

gvanrossum committed:
Fix a bug with CALL_STAT_INC

We were under-counting calls in `_PyEvalFramePushAndInit`
because the `CALL_STAT_INC` macro was redefined to a no-op
for the Tier 2 interpreter. The fix is a little convoluted.
This ought to result in ~37% more "Frames pushed" reported
under "Call stats". The new count is the correct one
(I presume).

mdboom reviewed (commented):
I understand how this is broken and why this change fixes it, so I'm marking as approved, but I agree it's "weird" / more convoluted than it needs to be.  Moving the order of functions in `ceval.c` would result in less weirdness (probably just moving `_PyEval_EvalFrameDefault` to the bottom since that's where all the macro updates happen) but that would create a lot of churn.

Alternatively, what if you rename `REAL_CALL_STAT_INC` to `CALL_STAT_INC_ALWAYS` and then just use `CALL_STAT_INC_ALWAYS` directly from `_PyEvalFramePushAndInit` (which I think is the only call site impacted by this change).  That would get rid of the weird dance of `#undef` and restoring just `CALL_STAT_INC` (but admittedly that would just replace it with another subtlety in `_PyEvalFramePushAndInit`).

gvanrossum committed:
Revert "Fix a bug with CALL_STAT_INC"

Let's try a different fix instead.

gvanrossum committed:
Move _PyEval_EvalFrameDefault to the end of the ceval.c

This should fix the issue with CALL_STAT_INC in a cleaner way
 (even if the diff is much larger).

gvanrossum committed:
Revert "Move _PyEval_EvalFrameDefault to the end of the ceval.c"

I'm going to try yet another approach.

gvanrossum committed:
Just don't redefine CALL_STAT_INC
